Tweak how the terminal is launched on Windows

fredrikekelund opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

## Related issues

- Fixes https://github.com/Automattic/studio/issues/359

## Proposed Changes

<del>Window distinguishes between user-level and system-level environment variables. Studio launches with the system-level `PATH` environment variable, which is then inherited by the command prompt when users click the `Open in… Terminal` button in Studio (this is just how Node's `child_process` module works).</del>

<del>This can be problematic for users because launching a new command prompt from the Start menu behaves differently. In that case, the user-level `PATH` environment variable is used. Starting a terminal with an unexpected `PATH` can obviously be troublesome since you may be missing the utilities you're used to.</del>

<del>This PR changes the logic for how the terminal is launched on Windows to fix this issue. The new logic depends on PowerShell. AFAIU, PowerShell [comes bundled](https://en.wikipedia.org/wiki/PowerShell) with all currently supported Windows versions, so this PR should hopefully work for essentially all of our users.</del>

<ins>After more testing, we concluded that we can't reliably reproduce this issue (see https://github.com/Automattic/studio/pull/814#issuecomment-2619381226). I repurposed this PR to focus on more cosmetic issues with how the terminal is launched on Windows</ins>

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Follow the instructions outlined in https://github.com/Automattic/studio/issues/359#issuecomment-2487958320 to install PHP and WP-CLI on Windows
2. Ensure that the user-level `PATH` envvar contains the WP-CLI path, but the system-level `PATH` does not (see [clarification](https://github.com/Automattic/studio/issues/359#issuecomment-2596106639))
3. Start Studio and have a Studio site
4. Click the `Open in… Terminal` button
5. Run `wp` to ensure that the default, user-level `PATH` envvar is loaded

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

fredrikekelund committed:
Load the default user-level environment variables in Windows terminal

fredrikekelund commented on src/ipc-handlers.ts:
I recommend viewing this diff with the `Hide whitespace` option enabled on GitHub, since a large part of my changes here is just re-indentation.

fredrikekelund commented on src/ipc-handlers.ts:
```diff
@@ -862,49 +862,71 @@ export async function getThumbnailData( _event: IpcMainInvokeEvent, id: string )
+	if ( platform === 'win32' ) {
+		// Windows
+		const isPowerShellAvailable = await checkPowerShellAvailability();
+		const defaultShell = process.env.ComSpec || 'cmd.exe';
```
This mirrors Node's internal logic for locating the default Windows shell. See https://nodejs.org/api/child_process.html#default-windows-shell

wojtekn commented:
I tested it on my Windows machine and confirmed that with the build from this PR, `wp` works in the Terminal session opened from Studio, the same as in the Terminal session opened outside Studio. After removing the wp-cli path from my user environment variable, both stopped working.

However, everything still worked fine when I tested the same with 1.3.1-beta1. 🤔

fredrikekelund commented:
I can confirm that in 1.3.2, Studio picks up the user-level `PATH` environment variable, and it's inherited correctly by the terminal. I'm researching what the behavior was like in v1.0.5 (since that's when the issue was reported).

fredrikekelund commented:
I can't reproduce it in v1.0.5, either… I'm beginning to wonder if this was only reproducible while running the development version of Studio to begin with

fredrikekelund committed:
Merge branch 'trunk' into f26d/load-default-terminal-envvars-on-windows

wojtekn commented:
@fredrikekelund do you mean development build? Or the app executed via `npm start`? If the latter, I can confirm I've never started Studio on Windows for development purposes.

fredrikekelund committed:
Simplified approach

fredrikekelund commented:
I meant when the app is started with `npm start`. 

> If the latter, I can confirm I've never started Studio on Windows for development purposes.

Interesting. Thanks for sharing that, @wojtekn.

fredrikekelund committed:
Enable Windows unit tests on CI

fredrikekelund committed:
Revert "Enable Windows unit tests on CI"

This reverts commit b2a6884780ccbbf55cdd5296df22ef375b9363ef.

fredrikekelund committed:
Restore macOS logic

fredrikekelund committed:
Tweaks

fredrikekelund committed:
Restore Linux logic

fredrikekelund commented:
Here's a quick **status update**. The tl;dr is that I've changed the scope of this PR to fix more cosmetic issues with how the terminal is launched on Windows.

- I can't reliably reproduce the issue that prompted this PR. AFAICT, it doesn't seem to affect built versions of Studio consistently. It's possible that the original problem was that we were modifying environment variables after Studio had launched, in which case they wouldn't get inherited by spawned shells.
- I chose to take a look at how GitHub desktop handles this problem (see [this directory](https://github.com/desktop/desktop/tree/development/app/src/lib/shells)) and ended up copying their solution for Windows.
  - Our original Windows solution is fundamentally very similar to theirs, but the benefit of theirs is that the title of the opened window is now `Command Prompt` instead of `C:\Windows\system32\cmd.exe` (which is the same as when you open the terminal from the Start menu).
  - I also modified the Linux solution to be more compact.

fredrikekelund committed:
Fix type error

wojtekn commented:
I tested the new build:
1. Ensured that wp-cli is in my user env variables
2. restarted system
3. started Studio and confirmed `wp` works in Terminal opened from Studio
